feat(kit): `Shimmer` add directive

mdlufy opened the pull request:
Fixes # <!-- link to a relevant issue. -->

mdlufy commented on projects/kit/directives/shimmer/shimmer.directive.ts:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    public ngOnChanges(): void {
+        if (!this.isBrowser) {
+            return;
+        }
```
Need to prevent the error `this.el.animate is not a function` on prerender build

mdlufy commented on projects/kit/directives/shimmer/shimmer.directive.ts:
```diff
@@ -0,0 +1,73 @@
+            return;
+        }
+
+        this.animation?.commitStyles();
+        this.animation?.cancel();
```
Fast switching `tuiShimmer` state force an error: Error: Uncaught (in promise): AbortError: The user aborted a request.

https://github.com/user-attachments/assets/f9b519df-5da3-4f2c-88da-6a7904779cfe

mdlufy replied:
Standart behaviour due to .cancel()
https://developer.mozilla.org/en-US/docs/Web/API/Animation/cancel#exceptions

waterplea commented on projects/kit/directives/shimmer/shimmer.directive.ts:
```diff
@@ -0,0 +1,83 @@
+
+    protected readonly nothing = tuiWithStyles(TuiShimmerStyles);
+
+    public shimmer = signal(false);
```
No need for it to be a signal

mdlufy replied:
Done

waterplea commented on projects/demo/src/modules/directives/shimmer/index.html:
```diff
@@ -0,0 +1,104 @@
+            visually communicate that content is loading, rather than missing or frozen.
+        </p>
+
+        <h3>How It Works</h3>
```
How it works is unnecessary. Instead let's add to the previous bullet that Shimmer is used when you have cached data that is currently being refreshed and if you have no data at all — a better choice would be Skeleton.

mdlufy replied:
Done

waterplea reviewed (changes requested):
Disabled animations example is wrong: what I meant was we need to react to --tui-duration. If it is set to 0 — we need to not apply the animations (both programmatical one and the CSS one) and instead just set disabled opacity. Also we need to disable pointer events, I think, and add attr.inert (nevermind browser support, it would be graceful degradation).

mdlufy commented on projects/demo/src/modules/directives/shimmer/examples/import/template.md:
````diff
@@ -0,0 +1,13 @@
+```html
+<!-- Usual HTML element -->
+<hgroup
````
Okay, updated example on usual HTML element.

I think it's important to show that shimmer can be used not only on regular elements, but also on other components of the Taiga UI. Therefore, I suggest leaving the second example with the calendar. 

P.S. this file will be included to llms-full.txt and the model will be able to explicitly answer: "Is it possible to use shimmer on other Taiga UI components" and show a minimal example.

mdlufy commented:
> Disabled animations example is wrong: what I meant was we need to react to --tui-duration. If it is set to 0 — we need to not apply the animations (both programmatical one and the CSS one) and instead just set disabled opacity. Also we need to disable pointer events, I think, and add attr.inert when shimmering (never mind browser support, it would be graceful degradation).

Updated, added react to `--tui-duration` value. @waterplea, will that be enough or we need to dynamically track `--tui-duration` value if it changes (via mutation observer, e.g.)?

waterplea commented on projects/kit/directives/shimmer/shimmer.directive.ts:
```diff
@@ -0,0 +1,83 @@
+                .then(() => {
+                    this.el.style.opacity = '';
+                })
+                .catch((error: unknown) => {
+                    // fast switching animation state force AbortError
+                    // https://developer.mozilla.org/en-US/docs/Web/API/Animation/cancel#exceptions
+                    if (error instanceof DOMException && error.name === 'AbortError') {
+                        return;
+                    }
+
+                    console.error(error);
+                });
```
Let's just change it to empty `catch`, no need to dig into the error.

waterplea commented on projects/demo/src/modules/directives/shimmer/examples/2/index.ts:
```diff
@@ -0,0 +1,40 @@
+    protected loading = true;
+
+    public avatars = [
+        {src: 'https://avatars.githubusercontent.com/mdlufy'},
```
Why not just strings here and in the first example?

waterplea commented on projects/demo/src/modules/directives/shimmer/examples/1/index.ts:
```diff
@@ -0,0 +1,47 @@
+export default class Example {
+    protected loading = true;
+
+    public avatars = [
```
protected readonly?

waterplea commented on projects/kit/directives/shimmer/shimmer.directive.ts:
```diff
@@ -0,0 +1,83 @@
+        }
+
+        this.disabled =
+            parseFloat(
```
Just `!parseFloat(...)`

waterplea commented on projects/kit/directives/shimmer/shimmer.directive.ts:
```diff
@@ -0,0 +1,83 @@
+
+        this.disabled =
+            parseFloat(
+                getComputedStyle(this.el).getPropertyValue('--tui-duration').trim(),
```
`trim()` is unnecessary, `parseFloat('     200ms   ')` works fine.
